### PR TITLE
feat: Contract read & write

### LINKED ISSUE #254: Smart Contract interaction: read
**Acceptance criteria**:
- A new tab should be displayed called "Read Contract"
- The tab should offer interaction with the smart contract entrypoints
- All sections should be expandable
- A Query button should be visible
- The output should be visible (including errors), under the Query button

Entrypoints should come from the smart contract verification while the interaction itself is done via the SDK.

### LINKED ISSUE #256: Smart contract interaction: write
Acceptance criteria:

- A new tab should be displayed called "Write Contract"
- The tab should offer interaction with the smart contract entrypoints
- All sections should be expandable
- A message should be displayed if the wallet is not connected. It's needed to interact with write entrypoints.
- A Write button should be visible
- The output should be visible (including errors), under the Query button

Entrypoints should come from the smart contract verification while the interaction itself is done via the SDK.

Dependency: Wallet Connection

### PR DESCRIPTION
**Pull request by janmichek**
## Description
resolves #254
resolves #256

## Demo

https://github.com/aeternity/aescan/assets/15363559/8a126508-9312-4065-b6be-54028085c8ac

## Checklist:

- [x] I have read and followed the [Contributing Guide](../../CONTRIBUTING.md)  
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by janmichek** [comment]
Testing notes for @Liubov-crypto

    go to https://studio.aepps.com/ and deploy some contracts to testnet
    do the verification in this PR preview on testnet. Select 7.4.0
    go to verified contract detail, now you can access read actions
    write actions cannot be verified, given the current limitations

**Comment by janmichek** [comment]
@Liubov-crypto And I found a way to test the write actions. Can you create new Superhero wallet account and send me the seed?

**Comment by Liubov-crypto** [comment]
I was getting an issue when tried to verify the smart contract. 
For example I've used: 
ct_svN4AGuUh2EJK4WfFQvY39Vpzd7awqmEZiFMybNG9t9gaSp98


ver

**Comment by janmichek** [comment]
> I was getting an issue when tried to verify the smart contract. For example I've used: ct_svN4AGuUh2EJK4WfFQvY39Vpzd7awqmEZiFMybNG9t9gaSp98
> 
> ver

What was the code of this contract? Is it some sample contract from aestudio?

**Comment by janmichek** [comment]
> > https://studio.aepps.com/
> 
> yes, from aestudio

None of the contracts from aestudio cannot be verified or just some of them?

**Comment by Liubov-crypto** [comment]
> > > https://studio.aepps.com/
> > 
> > 
> > yes, from aestudio
> 
> None of the contracts from aestudio cannot be verified or just some of them?

I tried a couple but was not working for me

**Comment by janmichek** [comment]
Here I deployed some contracts:
I deployed the contracts on behalf of your account. (That's why I needed the seed phrase.)
That makes you deployer and owner of those contracts.
And that enables you to do write actions.


hamster
ct_v5cSYQvkaq5oSKK4BTNBMgptRJSUC57DYwSjgEdGP5oi6Y7ia 
ct_JHVDvxUwSNiZDiMEqpjSF15gwPVvqRz2xXqf2vqw8aD76noSf 

AeForUsers
ct_22MwK52FbUYi82fQoCBP5ZVU2H8fsY5Ujg8YU356o7BuwLTbY 
ct_vtmQYHDzdxGigZD6VeXSgu2Df1XviBSVDRVssZz9bBYVKkXbf 

AeUnlockOnTime
ct_p8pcM6H8rdMnHUesw9fskRCFSCs428jthdkwjYyf8sPwr8oxp 
ct_vwDoPiv8nohvvW6gER1VrEUqaw21U9Mw4xJ2dmXYxQDAvnAcw 



scenario:
Only test testnet of this feature.

Switch the wallet for the account to which you send me the seed

Verify contract (ae studio examples + contract ids above)

goto verified contract detail

go to Write tab

connect wallet, follow instructions

execute write actions


- Another testing scenario is that if you want to access Write actions with connected wallet of the user who is NOT deployer of the contracts, then the Write actions should not be accessible (UI is grayed out) @Liubov-crypto

**Comment by Liubov-crypto** [comment]
Finally I've managed to verify at least one contract from the list above: ct_JHVDvxUwSNiZDiMEqpjSF15gwPVvqRz2xXqf2vqw8aD76noSf 

but when I followed steps attached in video I got an issue: the page became grey and some issues in console, @janmichek could you check it please: 


https://github.com/user-attachments/assets/17ec2c5f-0e79-4d4e-839d-9e6c0c463b6a

**Comment by janmichek** [comment]
> Finally I've managed to verify at least one contract from the list above: ct_JHVDvxUwSNiZDiMEqpjSF15gwPVvqRz2xXqf2vqw8aD76noSf
> 
> but when I followed steps attached in video I got an issue: the page became grey and some issues in console, @janmichek could you check it please:
> 2024-08-13.6.11.12.mov
> 
>  
> 
> 

@Liubov-crypto please retry now

**Comment by Liubov-crypto** [comment]
@janmichek  There are some issues again: UnsupportedVersionError: Unsupported node version 7.1.0. Supported: >= 6.2.0 < 7.0.0, please check it


https://github.com/user-attachments/assets/33e17175-93d3-4f56-9c44-758c602809ad

err node

**Comment by janmichek** [comment]
> @janmichek There are some issues again: UnsupportedVersionError: Unsupported node version 7.1.0. Supported: >= 6.2.0 < 7.0.0, please check it
> 2024-08-15.4.12.18.mov
> 
> err node

Fixed

File: src/components/ContractWritePanel.vue [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+} = contractVerifiedStore
+
+const route = useRoute()
+
+const response = ref([])
+const loadingIndex = ref(null)
+
+const isDisabled = computed(() => {
+  const connectedAddress = walletSdk ? Object.keys(walletSdk._accounts.current)[0] : null
+  return !walletSdk || contractDetails.value.createdBy !== connectedAddress
```
Comment (davidyuk) [inline src/components/ContractWritePanel.vue:51]:
```suggestion
  return !walletSdk
```
This is not an obvious restriction, contract should manage permissions by itself

**Comment by davidyuk** [comment]
For some reason I can't call init function of ct_HzBhnwxXg2zsPAmpX6ppoJNA5q1oojn8ZL2AsWozRxuBFERWD on testnet
Screenshot 2024-08-28 at 13 06 57

**Comment by davidyuk** [comment]
Does it make sense to accept arguments as JSON? I have a contract ct_24UQuk4JbCY2QhxnWvG257CBpirCvoCudXgLFj1MUaedKovUuX on testnet and I can't call its function currently
Screenshot 2024-08-28 at 13 10 04

Consider to combine "read contract" and "write contract" tabs.

**Comment by janmichek** [comment]
> ct_HzBhnwxXg2zsPAmpX6ppoJNA5q1oojn8ZL2AsWozRxuBFERWD

Intereseting, I didn't find such a contract on testnet or mainnet. Could you send me the source code, please? I will verify it

File: src/components/ContractWritePanel.vue (hunk 1 above)
Comment (janmichek) [reply src/components/ContractWritePanel.vue:51]: right, fixed

**Comment by janmichek** [comment]
> For some reason I can't call init function of ct_HzBhnwxXg2zsPAmpX6ppoJNA5q1oojn8ZL2AsWozRxuBFERWD on testnet Screenshot 2024-08-28 at 13 06 57

Yeah, I think that the init function should not be on the list at all. I filtered it out

**Comment by Liubov-crypto** [comment]
1. I've verified ct_JHVDvxUwSNiZDiMEqpjSF15gwPVvqRz2xXqf2vqw8aD76noSf and I've putted 11 and 22 into add_test_value 
and  2 and 3 into statefully_add_two. 

when I rechecked it on Read contract I got in read_test_value 5 as sum 2 and 3, and 33 in locally_add_two. are these numbers correct for the following fields? for me it looks like they switched. 

https://github.com/user-attachments/assets/22dc280e-8a38-4b74-853c-e572f8a9eb6b

2. also I've verified https://pr-872-aescan-testnet.stg.service.aepps.com/contracts/ct_p8pcM6H8rdMnHUesw9fskRCFSCs428jthdkwjYyf8sPwr8oxp?type=contract-write but can't see read contract fields, also new fields on write contract. is it correct? 

https://github.com/user-attachments/assets/51200f60-3c8c-4420-af7a-92186f942af3

3. when I switched between contract tabs wallet got disconnected.

**Comment by janmichek** [comment]
> ct I got in read_test_value 5 as sum 2 and 3, and 33 in locally_add_two. are these numbers correct for the following fields? for me it looks like they switched.

Thanks for the review 
1. It's correct like that. It depends on the order of the actions you run.
2. also correct, this contract has only writeable entrypoints
3. interesting. I cannot replicate the issue. Can you send a video to me, please?

**Review by Liubov-crypto — APPROVED** [review]
I was not able to reproduce ad 3 today. 

The rest is LGTM

**Pull request merged**